### ISSUE TITLE
Stop relying on Flask-OAuthlib

### ISSUE THREAD

**Issue by Sinclert** [inveniosoftware/invenio-oauthclient#195]
The package [Flask-OAuthlib](https://github.com/lepture/flask-oauthlib) has been discontinued and replaced by `authlib` (claimed both in the GitHub repository, and in [its docs](https://flask-oauthlib.readthedocs.io/en/latest/)).

Additionally, the usage of Flask-OAuthlib with the syntax `Flask-OAuthlib >= 0.9.3`, as it specified within the [setup.py](https://github.com/inveniosoftware/invenio-oauthclient/blob/master/setup.py#L74) at this very moment, leads to inconsistent `oauthlib` versions being install

### Inconsistency tracking

Let's track depedencies down, shall we?

`Flask-OAuthlib >= 0.9.3` resolves into `v0.9.5` ([last version in Pypi](https://pypi.org/project/Flask-OAuthlib/#history)), which declares the following ([check here](https://github.com/lepture/flask-oauthlib/blob/50f0924153859e6a76eec120ec3700ee69bde226/setup.py#L46)):
-  `oauthlib >= 1.1.2, ... <3.0.0` 👀 
- `requests-oauthlib >= 0.6.2`, which ends up as `requests-oauthlib v1.3.0` (latest version), with the following declaration ([check here](https://github.com/requests/requests-oauthlib/blob/792f1849089f87da59cb29118053f6d6f1bd199c/setup.py#L46)):
    - `oauthlib>=3.0.0` 👀

**Comment by lnielsen** [comment]
Replacing Flask-OAuthlib is quite big task as it's both in Invenio-OAuthClient and Invenio-OAuth2Server. Also, authlib is, like Flask-OAuthlib, a one-man show by the same person and I must admit that I'm quite concerned of spending a lot of effort migrating to authlib if he makes the same move again.

Obviously there's a problem here, just not clear what the solution is - i.e stay and fork flask-oauthlib, or switch to authlib, or …?

**Comment by Sinclert** [comment]
That's a good point.

Maybe for now just **locking** the `Flask-OAuthlib` version to  `0.9.3` or `0.9.4` is enough. It is just the latest version (`0.9.5`) the one that does not play well with `requests-oauthlib`.

Both `0.9.3` ([check here](https://github.com/lepture/flask-oauthlib/blob/9b407ff3883c2a25fc5053a1df5bf7c8f8115e7c/setup.py#L46)) and `0.9.4` ([check here](https://github.com/lepture/flask-oauthlib/blob/fabf71db84192a24fe7cb1b4409683bc53c229f6/setup.py#L46)) do not bound `oauthlib` with the "_<3.0.0_" bomb.

**Comment by slint** [comment]
In the `setup.py` we also pin `requests-oauthlib` to `<1.2.0`:

https://github.com/inveniosoftware/invenio-oauthclient/blob/ca0e40e6483d2f8fe8ccda670c56bc6423bb69b0/setup.py#L72 

...which should be enough to protect anyone doing `pip install invenio-oauthclient`.

The actual issue might be that `pip` messes up the installation order of the packages, which can indeed cause problems, e.g. if `Flask-OAuthlib` is installed first, what @Sinclert described above happens... 

Unfortunately, this is not an easy issue to fix, and that's also why when deploying Python apps in production, people have `requirements.txt` with the exact combination of pinned dependencies that are known to work, or use tools like `pipenv`, `poetry` and `dephell` exist.

Also, I guess `Flask-OAuthlib` pins `requests-oauthlib` for a reason (e.g. incompatibility)

**Comment by Sinclert** [comment]
You are right, `invenio-oauthclient` specifies `requests-oauthlib...<1.2.0`, which would solve the problem if the order is appropriate.

However, given that anyone who uses _requirements.txt_ is likely to specify dependencies in alphabetical order, `Flask-Oauthlib` will most certainly be installed before `invenio-oauthclient` for all _requirements.txt_ users, creating the incompatibility.

Could you limit `Flask-Oauthlib` in that regard?

Something like:
```python
    'Flask-OAuthlib>=0.9.3,<0.9.5',
```

**Comment by alexsdutton** [comment]
c.f. https://github.com/inveniosoftware/invenio-oauth2server/issues/195

**Comment by mfenner** [comment]
Migration to authlib would  make it easier to support OIDC, e.g. via OIDC Discovery. This would simplify adding support for additional OIDC providers. I am willing to work on this using the Pocket ID OIDC provider as an example use case.
